lib/Conversion/AffineToPipeline/AffineToPipeline.cpp

```diff
@@ -397,48 +431,101 @@ LogicalResult AffineToPipeline::createPipelinePipeline(
+                     pipeline.getStagesBlock().getArgument(i));
 
   // Create the stages.
   Block &stagesBlock = pipeline.getStagesBlock();
   builder.setInsertionPointToStart(&stagesBlock);
 
   // Iterate in order of the start times.
   SmallVector<unsigned> startTimes;
-  for (auto group : startGroups)
+  for (const auto &group : startGroups)
     startTimes.push_back(group.first);
   llvm::sort(startTimes);
 
   DominanceInfo dom(getOperation());
+  SmallVector<DenseSet<Value>> registerValues;
+  SmallVector<SmallVector<mlir::Type>> registerTypes;
+  SmallVector<BlockAndValueMapping> valueMaps;
+  DenseMap<Operation *, std::pair<unsigned, unsigned>> pipeTimes;
   for (auto startTime : startTimes) {
     auto group = startGroups[startTime];
-    OpBuilder::InsertionGuard g(builder);
 
     // Collect the return types for this stage. Operations whose results are not
     // used within this stage are returned.
     auto isLoopTerminator = [forOp](Operation *op) {
       return isa<AffineYieldOp>(op) && op->getParentOp() == forOp;
     };
-    SmallVector<Type> stageTypes;
-    DenseSet<Operation *> opsWithReturns;
+
     for (auto *op : group) {
+      if (op->getUsers().empty())
+        continue;
+      unsigned pipeStartTime = -1;
+      unsigned pipeEndTime = 0;
       for (auto *user : op->getUsers()) {
         if (*problem.getStartTime(user) > startTime || isLoopTerminator(user)) {
-          opsWithReturns.insert(op);
-          stageTypes.append(op->getResultTypes().begin(),
-                            op->getResultTypes().end());
+          pipeEndTime = std::max(pipeEndTime, *problem.getStartTime(user));
+          pipeStartTime = std::min(pipeStartTime, *problem.getStartTime(user));
         }
       }
+      // Skip iter if the users are in the same stage
+      if (pipeStartTime > pipeEndTime)
+        continue;
+
+      pipeTimes[op] = std::pair(pipeStartTime, pipeEndTime);
+
+      // Add register stages for each time slice
+      for (unsigned i = registerValues.size(); i <= pipeEndTime; ++i)
+        registerValues.push_back(DenseSet<Value>());
+
+      for (auto result : op->getResults())
+        registerValues.data()[startTime].insert(result);
+
+      // Handling multi-cycle ops here
+      for (unsigned i = std::max(startTime + 1, pipeStartTime); i < pipeEndTime;
+           i++) {
+        for (auto result : op->getResults())
+          registerValues.data()[i].insert(result);
+      }
     }
+  }
 
+  for (auto iterArg : innerLoop.getLoopBody().getArguments()) {
+    unsigned iterArgPipeLen = 0;
+    for (auto *user : iterArg.getUsers())
+      iterArgPipeLen = std::max(iterArgPipeLen, *problem.getStartTime(user));
+
+    for (unsigned i = 0; i < iterArgPipeLen; i++)
+      registerValues.data()[i].insert(iterArg);
+  }
+
+  // Now make register Types and valueMaps
+  for (unsigned i = 0; i < registerValues.size(); i++) {
+    SmallVector<mlir::Type> types;
+    for (auto val : registerValues.data()[i]) {
```

Comment:
Because you iterate over this here, I would use a SmallVector. You never use `contains` so you don't need a set: a vector will be cheaper. Also, it will be deterministic, so the same input IR will always produce the same output IR.